viewer/event_db.cpp

```diff
@@ -1,38 +1,122 @@
 #include "event_db.hpp"
 
+#include <algorithm>
+#include <queue>
+
 namespace HawkTracer
 {
 namespace viewer
 {
 
 EventDB::EventDB()
 {
-
 }
 
 void EventDB::insert(parser::Event event)
 {
-    _events.push_back(std::move(event));
+
+    auto klass_id = event.get_klass()->get_id();
+    if (_events.find(klass_id) == _events.end())
+    {
+        _events[klass_id] = std::vector<parser::Event>(1, std::move(event));
+    }
+    else
+    {
+        std::vector<parser::Event>::reverse_iterator it;
+        for (it = _events[klass_id].rbegin(); it != _events[klass_id].rend(); ++it)
+        {   
+            if ((*it).get_timestamp() < event.get_timestamp())
+            {
+                break;
+            } 
+        }
+        _events[klass_id].insert(it.base(), std::move(event));
+    }
+}
+
+void inplace_merge_events(std::vector<EventRef>& response, std::vector<int>& length, int l, int r)
+{
+    if (l == r)
+    {
+        return;
+    }
+
+    int offset_begin = l ? length[l - 1] : 0;
+    int offset_end = length[r];
+    int offset_middle = (offset_begin + offset_end) >> 1;
+
+    if (l + 1 == r)
+    {
+        std::inplace_merge(response.begin() + offset_begin, response.begin() + offset_middle, response.begin() + offset_end,
+                [](const EventRef& e1, const EventRef& e2) {
+                return e1.get().get_timestamp() < e2.get().get_timestamp();
+                });
+        return;
+    }
+
+    int m = (l + r) >> 1;
+    inplace_merge_events(response, length, l, m);
+    inplace_merge_events(response, length, m + 1, r);
+
+    std::inplace_merge(response.begin() + offset_begin, response.begin() + offset_middle, response.begin() + offset_end,
+            [](const EventRef& e1, const EventRef& e2) {
+            return e1.get().get_timestamp() < e2.get().get_timestamp(); 
+            });
 }
 
 std::vector<EventRef> EventDB::get_data(HT_TimestampNs start_ts, HT_TimestampNs stop_ts, const Query& query)
 {
-    std::vector<EventRef> events;
+    std::vector<EventRef> response;
+
+    // Dummy events for calling lower_bound
+    DummyEvent dummy_start_event(start_ts);
+    DummyEvent dummy_stop_event(stop_ts);
 
-    for (const auto& event : _events)
+    if (query.klass_id == (HT_EventKlassId)-1)
     {
-        auto ts = event.get_timestamp();
-        bool get_id_ok = query.klass_id == (HT_EventKlassId)-1 || event.get_klass()->get_id() == query.klass_id;
-        bool ts_ok = ts >= start_ts && ts <= stop_ts;
+        // Query all the event klasses
+        std::vector<int> lengths(_events.size(), 0);
+        int klass_index = -1;
```

Comment:
if you use the variable as an index of an array, it shouldn't be `signed int`, as it doesn't make sense to have negative index. Usually it should be `size_t`. However, I don't think you need this function at all, you could just insert values to `lengths` using `push_back` method. If you want to pre-allocate memory in the vector, you can call `vector::reserve()` method.